src/System.Text.RegularExpressions/src/System/Buffers/CharOwnedMemory.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Buffers
+{
+    internal sealed class CharOwnedMemory : OwnedMemory<char>
```

Comment:
I do not understand why this is needed. Why can't you just pass the Span or pinned pointer around (wrapped in a struct so that it does not look ugly)?

Reply:
This is only used for Regex.Replace with a MatchEvaluator overload. As the evaluator can reference any public member of the Match object it could reference  the Value or the ValueMemory. As I store input string in the Match object as a ReadOnlyMemory<char> (https://github.com/ViktorHofer/corefx/pull/1/files#diff-49b12395ba18f98f80686a1e251ab22cR46) the Span needs to be converted. I don't see any other way here without allocating the text in a string.

Reply:
Why is ValueMemory a memory? Why it is not a Span instead?

Reply:
Because 1) the Match object is returned to caller and 2) the type itself isn't a byref type, I can't store the Span in there, even if I want to.

Reply:
Yes, you would have to store a pinned pointer in there. It is what you are doing already - I do not see why you need an extra level of indirection to store a pinned pointer.

Reply:
If you are doing this for security - it does not make things any more secure. You can achieve about the same level of security without CharOwnedMemory.

To make things actually secure, you would need to introduce a new `ref struct MatchValue` that store the Span in just fine, and that the C# language guarantees does not leak.

Reply:
But this only applies to the Replace call with the MatchEvaluator? You aren't talking about using a pinned pointer even for the Match & Matches calls where I return the object back to the caller, right? I had a discussion with Stephen about that offline and we both agreed that pinning a pointer of the lifetime of the call is risky / not even feasible?

Reply:
over the lifetime, not of.

Reply:
I think that instead of CharOwnedMemory  - that is relative expensive because of it is an object with finalizer. You can have a struct like this instead:

```
struct MemoryOrPinnedSpan
{
    Memory<char> _memory;
    char* _pinnedSpan;
    int _length;

    Span<char> AsSpan()
    {
        return (_pinnedSpan != null) ? new Span(_pinnedSpan, _length) : _memory.AsSpan();
    }
}
```

Reply:
and with that you also propose to change the exposed ValueMemory to ValueSpan? I just want to be clear that we are talking here about all API calls not just about the special Replace case.

Reply:
Yes.

Also, my preference would be to have the API secure-by-design so that it is impossible to get dangling pointers. It would mean to introduce a new `ref struct MatchValue`

Reply:
Talked with Jan offline, will incorporate his feedback and will put additional API review feedback in a remarks section at the top.

Reply:
API calls which are unrelated to the Span pinning (everything except Replace with MatchEvaluator) "suffer" under the additional wrapper which introduces two additional sizeof(int) members. Not really painful but worth to point out...